datafusion/common/src/tree_node.rs

```diff
@@ -204,27 +176,64 @@ pub trait TreeNode: Sized {
     ///
     /// The nodes are visited using the following order:
     /// ```text
-    /// f_down(ParentNode)
-    /// f_down(ChildNode1)
-    /// f_up(ChildNode1)
-    /// f_down(ChildNode2)
-    /// f_up(ChildNode2)
-    /// f_up(ParentNode)
+    /// TreeNodeRewriter::f_down(ParentNode)
+    /// TreeNodeRewriter::f_down(ChildNode1)
+    /// TreeNodeRewriter::f_up(ChildNode1)
+    /// TreeNodeRewriter::f_down(ChildNode2)
+    /// TreeNodeRewriter::f_up(ChildNode2)
+    /// TreeNodeRewriter::f_up(ParentNode)
     /// ```
     ///
     /// See [`TreeNodeRecursion`] for more details on how the traversal can be controlled.
     ///
-    /// If `f_down` or `f_up` returns [`Err`], recursion is stopped immediately.
```

Comment:
Do we need to remove this? I really need this and I would rather keep this here than adding it back in the next PR.

Reply:
If you don't like that we reuse the name `transform` then we can rename it like `transform_down_up`.

Reply:
Instead of introducing a new API, could we consider using `transform_down` followed by `transform_up`? I'm not trying to make things difficult for you, but I'm just curious if we could achieve the same result by adding an extra line of code where needed.

Reply:
No, that's not the same. I will need a short form of `rewrite()` (top-down and bottom-up closures provided for one combined traversal). This is an example how I would like to use such an API:
https://github.com/apache/arrow-datafusion/pull/8891#issuecomment-1933706262

Reply:
I got your point. Not related with the topic,  but I guess `contains_pattern` in your example deep dives on each node until the leaves, so adding such an `f_down` may get worse the performance actually.

However, in general there may exist such use cases where f_down can speed up the f_up phase. I'm reverting this change. Can I just ask from you to modify the transform adapting the same behavior with updated visit and rewrite? I have updated the test results, but cannot adapt the transform implementation yet.

Reply:
I think calculating the pattern bitset will be very cheap. (I want to do that lazyly). Also, I think in many cases several subtrees remain unchanged during transformations and the patterns of those subtrees don't need to be recalculated, which means that multiple subsequent transformations can take advantage of some patterns calculated during the first transformation.

> I'm reverting this change. Can I just ask from you to modify the transform adapting the same behavior with updated visit and rewrite?

Thanks. Sure, I will do that. (Most likely only tomorrow...)